### ISSUE TITLE
Add support for computing the cumulative sum to the standard

### ISSUE THREAD

**Issue by steff456** [data-apis/array-api#597]
This RFC requests to include a new API in the array API specification for the purpose of computing the cumulative sum.

## Overview

Based on array comparison [data](https://github.com/data-apis/array-api-comparison/blob/7ecfa9867ec43a28e914b35868c041443d31d772/data/common_apis.csv#L43), the API is available in all the libraries in the PyData ecosystem.

## Prior art

- NumPy: <https://numpy.org/doc/stable/reference/generated/numpy.cumsum.html>
- PyTorch: https://pytorch.org/docs/stable/generated/torch.cumsum.html
- MXNet: <https://mxnet.apache.org/versions/1.6/api/r/docs/api/mx.nd.cumsum.html>
- TensorFlow: <https://www.tensorflow.org/api_docs/python/tf/math/cumsum>

## Proposal:

```python
def cumsum(x: array, /, *, axis: Optional[Union[int, Tuple[int, ...]]] = None, dtype: Optional[dtype] = None) -> array
```

- `dtype` kwarg is for consistency with `sum` et al

cc @oleksandr-pavlyk

**Comment by asmeurer** [comment]
There's also `np.add.accumulate` which may be relevant when looking at the comparison data.

**Comment by rgommers** [comment]
Based on previous experience with complaints about bad naming (see, e.g., https://github.com/scipy/scipy/issues/12924 and https://www.reddit.com/r/programminghorror/comments/j6sd61/i_was_just_looking_at_the_documentation_for/) I would very much prefer not to enshrine `cumsum` and `cumprod` (as proposed in gh-598) names in this API standard.

This is pretty niche functionality and arguably not "core" enough to implementing an array library for it to be in this standard at all, so I'd vote for leaving it out completely. In a compat layer it could perhaps be named `cumulative_*` like SciPy did with a few methods, if it's desired for these functions to be there.

**Comment by soraros** [comment]
Niche it may be, I still think `cumsum` is pretty useful, as many indexing tricks depend on it. This is especially true when one works with statically-shaped system like JAX where such tricks are more or less required. For instance, its usages in the implementation of [`jax.numpy.nonzero`](https://github.com/google/jax/blob/3838d7612a249264d7e3b89eb96c3e91eb71f9a9/jax/_src/numpy/lax_numpy.py#L1376) and [`jax.numpy.repeat`](https://github.com/google/jax/blob/3838d7612a249264d7e3b89eb96c3e91eb71f9a9/jax/_src/numpy/lax_numpy.py#L2623-L2632) are pretty typical.

**Comment by rgommers** [comment]
> This is especially true when one works with statically-shaped system like JAX where these tricks are more or less required.

@soraros it seems like all that usage of `cumsum` in JAX is just a cumbersome way of working around not having boolean indexing support for in-place ops? And JAX should instead add some other primitive that's more suitable, rather than letting all users manually construct expanded integer index arrays to then use with `.at[idx_cumsum].xxx`?

```python
# For an array `xs` and a boolean index `conds` of the same space
# example JAX expression for `jax_filter`:
>>> cumsum = jnp.cumsum(conds)
>>> jnp.zeros_like(xs).at[cumsum - 1].add(jnp.where(conds, xs, 0))

# NumPy:
>>> xs = np.arange(5)
>>> conds = np.array([True, False, True, False, True])
>>> ys = np.zeros_like(xs)
>>> ys += np.where(conds, xs, 0)
>>> ys
array([0, 0, 2, 0, 4])
>>> xs[conda]
array([0, 2, 4])
```

That kind of JAX code looks very bad, and as a motivator for writing portable code based on a standard I don't think it's a positive. JAX could add support for `x[ix_bool] += 1` tomorrow, and imho they should rather than make users write `x.at[cumsum(ix_bool)].add(1)`. It's 100% equivalent. And for the "expanded boolean indexing" where the desired result is `[0, 0, 2, 0, 4]` as an output, it needs a new builtin function like `jnp.bool_index(xs, conds)`. It has little to do with cumulative sums for data/statistical purposes, it's a type of indexing.

There's a known blocker (brought up by the JAX team before) for adding more in-place operator support beyond what they have now, which is that this standard should be able to better guarantee that `+=` doesn't modify views in NumPy et al. But if we can make that guarantee at some point, then we have what we want here.

**Comment by soraros** [comment]
> @soraros it seems like all that usage of `cumsum` in JAX is just a cumbersome way of working around not having boolean indexing support for in-place ops? 

@rgommers Yes and no. It is a cumbersome workaround, but I also think the problem is more fundamental than that. JAX is essentially a front-end for XLA, and the primitives provided by XLA (for now) require static shape. So the line that actually go wrong is

```python
>>> xs[ix_bool]
array([0, 2, 4])
```

Note this code does work in JAX, though not jittable, for we don't know its output shape. Let's pretend `x[ix_bool] += 1` is syntax sugar for `x = x + where(ix_bool, 1, 0)` (which works in JAX) for a moment. The same problem appears when we want `x[ix_bool] += [1, 3, 5]`. Again, we somehow need to know the shape of the rhs, which is equivalent to know the shape of `xs[ix_bool]` as in the last example.

So what we really work around is the static shape requirement (recall the need of a `size` parameter for `nonzero`), which is not exclusively JAX.

<details>
<summary>Now, for something a bit off-topic.</summary>

I think the JAX style functional syntax `a = a.at[...].set(...)` for in-place operation looks (and arguably works) better than `numpy`, and I'd really like to have it for array api. Some pros:

- Looks familiar, and simulates the feel of in-place operation just fine. 
- Made it clear nothing is modified. This restricted access pattern would work with any accelerator-backed system. _I think_ it would aid static analysis in system like `Numba` as well.
- More concise, can be chained, and sometimes express our intention better.

```python
a = zeros(m)       # initialing a
a[I] += arange(n)  # semantically, still initialing a

# VS

# being concise here is not the important point
# this line becomes a "semantical block" for initialisation
a = zeros(m).at[I].add(arange(n))  # initialing a
```

- Can specify indexing mode, (more) easily.
```python
# I think these are fairly cumbersome to represent in `numpy`, as we don't have kwargs for __getitem__
b = a.at[I].add(val, unique_indices=True)     # important info for accelerators
c = b.at[J].get(mode='fill', fill_value=nan)  # sure, we have `take`, but this is uniform and cool
```

</details>

**Comment by rgommers** [comment]
@soraros thanks for all this detail, it's very interesting actually. I think there's something to be said indeed for `.at` - especially if JAX can make a new cleaner API for it that doesn't rely on things like explicit use of `cumsum`. This discussion is effectively a follow-up to gh-84. What do you think about opening a new issue to continue this discussion, especially your "for something a bit off-topic" part?

**Comment by soraros** [comment]
@rgommers Glad you find the exchange interesting!
[gh-84](https://github.com/data-apis/array-api/issues/84) is indeed interesting, I will read that throughly later.
I'm not sure how to proceed regarding opening a new issue to continue the discussion though (not sure about the exact topic and/or scope you have in mind). If it's not too much trouble, could you open a new issue and I can move my "off-topic" part there? Thanks in advance!

**Comment by asmeurer** [comment]
How do you implement cumulative sum using only array API functions (and without using a Python loop)?

**Comment by oleksandr-pavlyk** [comment]
If use of linear algebra is allowed:

```
In [4]: def cumsum(x):
   ...:     assert x.ndim == 1
   ...:     n = x.shape[0]
   ...:     return tril(ones((n,n,))) @ x
   ...:

In [5]: cumsum(np.array([1,2,3,4,5]))
Out[5]: array([ 1.,  3.,  6., 10., 15.])
```

**Comment by rgommers** [comment]
> I'm not sure how to proceed regarding opening a new issue to continue the discussion though (not sure about the exact topic and/or scope you have in mind). If it's not too much trouble, could you open a new issue so I (or maybe you) can move the relevant part there? Thanks in advance!

Done now in gh-609 - sorry for the delay! I spent a lot of time refreshing my memory and trying to put together something more coherent. But it's tricky; the need for new API in JAX to avoid `cumsum` seems clear, but if it was easy to exactly define how it'd look, the JAX devs would have done that by now I guess:)

**Comment by shoyer** [comment]
I think cumulative sum is a rather fundamental array operation and we should include it in the standard. None of the typical reasons for omitting a function from the API standard apply here:

- It has well defined behavior (aside from NumPy's funny flattening behavior if `axis` is omitted)
- It is not particularly hard to implement in a distributed fashion or on accelerators, as evidenced by how it can be found in Dask and JAX.
- It is not easy to implement in terms of other fundamental operations.

To give a few other examples of how I've used it:

1. Calculating integrals
2. Calculating moving window averages
3. Calculating cumulative probabilities 

As a reference point on popularity, I see about twice as many uses of `np.cumsum` in Google's codebase as for `np.roll` or `np.var`.

I'll also second Ralf's point on calling it `cumulative_sum` rather than `cumsum`.

**Comment by WarrenWeckesser** [comment]
> It has well defined behavior...

Before locking in the specification, it would be worthwhile taking a look at https://github.com/numpy/numpy/issues/6044, and https://github.com/numpy/numpy/pull/14542 that is referenced from https://github.com/numpy/numpy/issues/6044.  My [comment](https://github.com/numpy/numpy/pull/14542#issuecomment-586494608) in https://github.com/numpy/numpy/pull/14542 provides some evidence for the usefulness of allowing the result to include a prepended 0.  (More generally, for other cumulative operations such as `cumprod`, the identity of the operation would be prepended.)

**Comment by shoyer** [comment]
> Before locking in the specification, it would be worthwhile taking a look at [numpy/numpy#6044](https://github.com/numpy/numpy/issues/6044), and [numpy/numpy#14542](https://github.com/numpy/numpy/pull/14542) that is referenced from [numpy/numpy#6044](https://github.com/numpy/numpy/issues/6044). My [comment](https://github.com/numpy/numpy/pull/14542#issuecomment-586494608) in [numpy/numpy#14542](https://github.com/numpy/numpy/pull/14542) provides some evidence for the usefulness of allowing the result to include a prepended 0. (More generally, for other cumulative operations such as `cumprod`, the identity of the operation would be prepended.)

I agree, starting with zero (and excluding the last value) is quite useful, and I would definitely support adding an optional argument.

Is it clear that this is a better default behavior? My inclination is that it would not be worth the trouble to break existing code.

**Comment by WarrenWeckesser** [comment]
> Is it clear that this is a better default behavior? My inclination is that it would not be worth the trouble to break existing code.

I agree, starting with zero is not a better default behavior, if only because of the long history of the existing behavior in numpy.  The links to those previous discussions are more for increasing awareness of the interest and usefulness of this option.  That is, don't lock in an API that can't be enhanced with an option later. I don't think that is a problem here, but I don't know how hard it will be to add options to a function in the array API specification later. Also, maybe the reminder will inspire someone to push forward with the enhancement in numpy 🤞 or in other libraries.

**Comment by arogozhnikov** [comment]
Just opinion: 
- cumsum in my experience is very common, and I'd even vote for cumulative logsumexp (but not cumprod)
- Prepending zero - had a number of cases when this would be a desired behavior.

**Comment by rgommers** [comment]
Looks like there's enough thumbs-ups for cumulative sum at least, and we should add it now. @kgryte just volunteered to open a PR for it.

**Comment by seberg** [comment]
On the PR there is currently `startpoint=` and `endpoint=`, so I am wondering if anyone here has opinions on naming?  (Above was `start_with_zero=` also).
I don't want to include it here, but in principle it makes sense to have `initial=value` (distinct or included in a `start_with_initial=True`/`include_initial`).  I am wondering if there is a clearer name for `startpoint=True`?

**Comment by seberg** [comment]
On the NumPy issue https://github.com/numpy/numpy/issues/6044 `include_identity` and `include_final` were brought up.  I think I lean towards those names being nicer, `include_initial` would pair better with an `initial=` argument for reductions, but not sure it is as clear.

**Comment by kgryte** [comment]
I've updated the PR to use `include_initial`.
